### PR TITLE
Support listing multiple images to remove

### LINKED ISSUE #1190: Remove multiple images at once
Hi,
First of all, thanks for outstanding plugin, guys. 

It would be great if I could delete multiple docker images with a single Gradle task.
### Expected Behavior

Additional task ```DockerRemoveImages``` for example or upgrade current ```DockerRemoveImage``` to accept a list of image Ids.

### Current Behavior

There is no way to delete multiple images at once **or** I couldn't find it. Anyone please guide me if I'm missing something. Now I have to do something like that:
```Gradle
def imagesToClean = []

import com.bmuschko.gradle.docker.tasks.image.*
import com.github.dockerjava.api.command.RemoveImageCmd;
abstract class DockerRemoveImages extends DockerExistingImage {

    DockerRemoveImages () {
        imageId = ''
    }

    @Input
    public final SetProperty<String> getImageIds() {
        return imageIds
    }
    private final SetProperty<String> imageIds = project.objects.setProperty(String.class);

    @Override
    public void runRemoteCommand() {
        getImageIds().get().each {
            i -> 
                getLogger().quiet("Removing image with ID {}.", i);
                                
                RemoveImageCmd removeImageCmd = getDockerClient().removeImageCmd(i);
                removeImageCmd.withForce(Boolean.TRUE);                

                removeImageCmd.exec();                
        }
    }
}

tasks.register('listLocalImages', DockerListImages) {    
    imageName = DOCKER_IMAGE_NAME
    onNext { i -> imagesToClean.add(i.Id) }
}

tasks.register('cleanLocalImages', DockerRemoveImages) {
    dependsOn tasks.named('listLocalImages')
    
    imageIds = imagesToClean    
}
```
Sure, its a workaround. But...

### Context

On my build server there are a lot of Docker images left which were already pushed to remote registry. Would like to have an easy way to remove them as a part of Gradle script. Also, if Gradle script fails after image has been built its a good idea on next successful run remove any stale images. Sure, there could be more than just one.

### Your Environment

```Gradle
plugins {
    id 'com.bmuschko.docker-remote-api' version '9.3.2'
}
```
```
./gradlew --version

------------------------------------------------------------
Gradle 8.3
------------------------------------------------------------

Build time:   2023-08-17 07:06:47 UTC
Revision:     8afbf24b469158b714b36e84c6f4d4976c86fcd5

Kotlin:       1.9.0
Groovy:       3.0.17
Ant:          Apache Ant(TM) version 1.10.13 compiled on January 4 2023
JVM:          19.0.2 (Private Build 19.0.2+7-Ubuntu-0ubuntu322.04)
OS:           Linux 6.2.0-26-generic amd64
```

### PR DESCRIPTION
**Pull request by marcospereira**
## What?

Fixes #1190.

It allows users to list multiple images to remove.

## How?

A new `ListProperty` allows users to set a list of the images they want to remove. The old property is maintained, but it will generate a warning if used and throw an error if used together with the list property.

## Status

This is a WIP/draft for now. But I think it is close to be ready to merge.

### CODE REVIEW

File: .github/workflows/linux-build-release.yml [hunk 1]
```diff
@@ -38,6 +38,12 @@ jobs:
         run: ./gradlew functionalTest
       - name: Documentation tests
         run: ./gradlew docTest
+      - name: Upload test reports
```
Comment (bmuschko) [inline .github/workflows/linux-build-release.yml:41]: There's really no need to upload test results, as you can view them with the help of a build scan. The build scan link is available at the end of each workflow run, on success and on failure. I'd say remove this change.

File: .github/workflows/windows-build.yml [hunk 2]
```diff
@@ -23,5 +23,11 @@ jobs:
         run: ./gradlew test
       - name: Integration tests
         run: ./gradlew integrationTest
+      - name: Upload test reports
```
Comment (bmuschko) [inline .github/workflows/windows-build.yml:26]: There's really no need to upload test results, as you can view them with the help of a build scan. The build scan link is available at the end of each workflow run, on success and on failure. I'd say remove this change.

File: src/main/java/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImage.java [hunk 3]
```diff
@@ -16,29 +16,151 @@
...
+    @Input
+    @Optional
+    public final Property<Boolean> getNoPrune() {
+        return noPrune;
+    }
+
     private final Property<Boolean> force = getProject().getObjects().property(Boolean.class);
+    private final Property<Boolean> noPrune = getProject().getObjects().property(Boolean.class);
+
+    private final ListProperty<String> images = getProject().getObjects().listProperty(String.class);
```
Comment (bmuschko) [inline src/main/java/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImage.java:52]: Have a look at `DockerPushImage`. It does something similar.

File: src/main/java/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImage.java [hunk 4]
```diff
@@ -16,29 +16,151 @@
...
+
+    /**
+     * Sets the IDs or names of the images to be removed.
+     *
+     * @param images the names or IDs of the images.
+     * @see #images(ListProperty)
+     */
+    public void images(List<String> images) {
```
Comment (bmuschko) [inline src/main/java/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImage.java:60]: You will just need a method that returns a reference to the `images` field.

```
@Input
public final SetProperty<String> getImages() {
    return images;
}
```

File: src/main/java/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImage.java [hunk 5]
```diff
@@ -16,29 +16,151 @@
...
+     * Sets the IDs or names of the images to be removed.
+     *
+     * @param images Image ID or name as {@link Provider}
+     * @see #images(String...)
+     */
+    public void images(ListProperty<String> images) {
+        this.images.set(images);
+    }
+
+    // Overriding methods to provide a warning message.
```
Comment (bmuschko) [inline src/main/java/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImage.java:96]: Remove all of those methods and extend from `AbstractDockerRemoteApiTask` instead. The parent class `DockerExistingImage` is really only meant for a single image reference.

File: src/functTest/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImageFunctionalTest.groovy [hunk 6]
```diff
@@ -42,6 +42,75 @@ class DockerRemoveImageFunctionalTest extends AbstractGroovyDslFunctionalTest {
         !result.output.contains("repository")
     }
 
+    def "can remove multiple images"() {
```
Comment (bmuschko) [inline src/functTest/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImageFunctionalTest.groovy:45]: We'll also want to change all documentation and test cases to use the new method instead.

File: src/functTest/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImageFunctionalTest.groovy [hunk 7]
```diff
@@ -42,6 +42,75 @@ class DockerRemoveImageFunctionalTest extends AbstractGroovyDslFunctionalTest {
...
+                from '$thirdImage'
+                label(['maintainer': 'jane.doe@example.com'])
+            }
+            task buildThirdImage(type: DockerBuildImage) {
+                dependsOn thirdDockerfile
+                inputDir = file("build/docker")
+            }
+
+            task removeImages(type: DockerRemoveImage) {
+                dependsOn buildFirstImage
```
Comment (bmuschko) [inline src/functTest/groovy/com/bmuschko/gradle/docker/tasks/image/DockerRemoveImageFunctionalTest.groovy:86]: We won't really need the `dependsOn` as you set an implicit dependency by assigning the image references (the output of `DockerBuildImage` tasks) to the input field `images`.

**Comment by marcospereira** [comment]
Thanks for the review, @bmuschko. I got busy and haven't had the time to finish this, but your comments will definitely help to get this done later. :-)

**Comment by bmuschko** [comment]
I am going to close this issue. We can reopen if you or someone else is planning to continue working on this change.

**Pull request closed without merging**